src/compiler/emitter.ts

```diff
@@ -4337,16 +4369,38 @@ namespace ts {
 
         // Source Maps
 
+        function getParsedSourceMap(node: UnparsedSource) {
+            if (node.parsedSourceMap === undefined && node.sourceMapText !== undefined) {
+                node.parsedSourceMap = tryParseRawSourceMap(node.sourceMapText) || false;
+            }
+            return node.parsedSourceMap || undefined;
+        }
+
         function pipelineEmitWithSourceMap(hint: EmitHint, node: Node) {
             const pipelinePhase = getNextPipelinePhase(PipelinePhase.SourceMaps, node);
-            if (isUnparsedSource(node) && node.sourceMapText !== undefined) {
-                const parsed = tryParseRawSourceMap(node.sourceMapText);
+            if (isUnparsedSource(node)) {
+                const parsed = getParsedSourceMap(node);
                 if (parsed) {
                     sourceMapGenerator!.appendSourceMap(
                         writer.getLine(),
                         writer.getColumn(),
                         parsed,
-                        node.sourceMapPath!);
+                        node.sourceMapPath!,
+                        node.pos ? node.getLineAndCharacterOfPosition(node.pos) : undefined);
+                }
+                pipelinePhase(hint, node);
+            }
+            else if (isUnparsedPrologue(node)) {
+                const parsed = getParsedSourceMap(node.parent);
+                if (parsed) {
+                    sourceMapGenerator!.appendSourceMap(

```

Comment:
Generally speaking, `node.sourceMapText` existing implies that `sourceMapGenerator` should exist (since we only fill that field when it's needed), however if we get an `UnparsedSource` via a user-provided transform, that may not be the case. To be defensive, I'd check that `sourceMapGenerator` is defined, in addition to `parsed`, rather than using a nonnull assert here.